File: python/tests/test_ch23_accum.py
# region imports
import os
import sys
import logging
import pytest
import io

''' Notes
1. 
2.
'''
# from  pytest_mock import mocker
# endregion imports
# region globals
logger = logging.getLogger(__name__)
# endregion globals
# region helpers
# endregion helpers
# region tests for Ch23
def test_2321_map():
    lst = [2,5,9]
    lst3 = list(map(lambda x: 3 * x, lst))
    elst3 = [6, 15, 27]
    assert elst3 == lst3

def test_2321_lc():
    lst = [2,5,9]
    lst3 = [3*x for x in lst]
    elst3 = [6, 15, 27]
    assert elst3 == lst3

def test_2322_map():
    '''  Using map, create a list assigned to the variable greeting_doubled that doubles each element in the list lst.
    '''
    lst = [["hi", "bye"], "hello", "goodbye", [9, 2], 4]
    greeting_doubled = list(map(lambda x : 2 * x , lst))
    logger.info(greeting_doubled)
    elst = [["hi", "bye", "hi", "bye"], "hellohello", "goodbyegoodbye", [9, 2, 9, 2], 8]
    assert elst == greeting_doubled
def test_2322_lc(): # list comprehension version
    lst = [["hi", "bye"], "hello", "goodbye", [9, 2], 4]
    greeting_doubled = [ 2 * x for x in lst]
    logger.info(greeting_doubled)
    elst = [["hi", "bye", "hi", "bye"], "hellohello", "goodbyegoodbye", [9, 2, 9, 2], 8]
    assert elst == greeting_doubled

def test_2323_map():
    ''' Below, we have provided a list of strings called abbrevs. 
    Use map to produce a new list called abbrevs_upper that contains all the same strings in upper case.
    '''
    abbrevs = ["usa", "esp", "chn", "jpn", "mex", "can", "rus", "rsa", "jam"]
    abbrevs_upper = list(map(lambda s: s.upper(),abbrevs))
    # logger.info(abbrevs_upper)
    eabbrevs = ["USA", "ESP", "CHN", "JPN", "MEX", "CAN", "RUS", "RSA", "JAM"]
    assert eabbrevs == abbrevs_upper

def test_2323_lc(): 
    abbrevs = ["usa", "esp", "chn", "jpn", "mex", "can", "rus", "rsa", "jam"]
    abbrevs_upper = [ s.upper() for s in abbrevs]
    # logger.info(abbrevs_upper)
    eabbrevs = ["USA", "ESP", "CHN", "JPN", "MEX", "CAN", "RUS", "RSA", "JAM"]
    assert eabbrevs == abbrevs_upper

def test_2331_filter():
    lst = [3, 4, 6, 7, 0, 1]
    evens = list(filter(lambda x : x % 2 == 0,lst))
    elst = [4, 6, 0]
    assert elst == evens

def test_2331_lc():
    lst = [3, 4, 6, 7, 0, 1]
    evens = [x for x in lst if x % 2 == 0]
    elst = [4, 6, 0]
    assert elst == evens

def test_2332_filter():
    ''' Write code to assign to the variable filter_testing all the elements in lst_check
    that have a w in them using filter.
    '''
    lst_check = ['plums', 'watermelon', 'kiwi', 'strawberries', 'blueberries', 'peaches', 'apples', 'mangos', 'papaya']
    filter_testing = list(filter(lambda w : "w" in w,lst_check))
    elst = ['watermelon', 'kiwi', 'strawberries']
    assert elst == filter_testing

def test_2332_lc():
    lst_check = ['plums', 'watermelon', 'kiwi', 'strawberries', 'blueberries', 'peaches', 'apples', 'mangos', 'papaya']
    filter_testing = [w for w in lst_check if "w" in w]
    elst = ['watermelon', 'kiwi', 'strawberries']
    assert elst == filter_testing

def test_2341_mapfilter():
    things = [3, 4, 6, 7, 0, 1]
    #chaining together filter and map:
    # first, filter to keep only the even numbers
    # double each of them
    lst = list(map(lambda x: x*2, filter(lambda y: y % 2 == 0, things)))
    elst = [8, 12, 0]
    assert elst == lst

def test_2341_lc():
    things = [3, 4, 6, 7, 0, 1]
    lst = [x*2 for x in things if x % 2 == 0]
    elst = [8, 12, 0]
    assert elst == lst

def test_2342_dict_mapfilter():
    tester = {'info': [{"name": "Lauren", 'class standing': 'Junior', 'major': "Information Science"},{'name': 'Ayo', 'class standing': "Bachelor's", 'major': 'Information Science'}, {'name': 'Kathryn', 'class standing': 'Senior', 'major': 'Sociology'}, {'name': 'Nick', 'class standing': 'Junior', 'major': 'Computer Science'}, {'name': 'Gladys', 'class standing': 'Sophomore', 'major': 'History'}, {'name': 'Adam', 'major': 'Violin Performance', 'class standing': 'Senior'}]}
    compri = list(
        map( lambda item: item['name'], 
           filter(lambda item: 'name' in item , tester['info'])
        )
    )
    ecompri = ['Lauren', 'Ayo', 'Kathryn', 'Nick', 'Gladys', 'Adam']
    assert ecompri == compri

def test_2342_dict_lc():
    ''' Write code to assign to the variable compri all the values of the key name in any of the sub-dictionaries in the dictionary tester. 
    Do this using a list comprehension.
    '''
    tester = {'info': [
        {"name": "Lauren", 'class standing': 'Junior', 'major': "Information Science"},
        {'name': 'Ayo', 'class standing': "Bachelor's", 'major': 'Information Science'}, 
        {'name': 'Kathryn', 'class standing': 'Senior', 'major': 'Sociology'}, 
        {'name': 'Nick', 'class standing': 'Junior', 'major': 'Computer Science'}, 
        {'name': 'Gladys', 'class standing': 'Sophomore', 'major': 'History'}, 
        {'name': 'Adam', 'major': 'Violin Performance', 'class standing': 'Senior'}]}
    compri = [item['name'] for item in tester['info'] if 'name' in item]
    logger.info(compri)
    ecompri = ['Lauren', 'Ayo', 'Kathryn', 'Nick', 'Gladys', 'Adam']
    assert ecompri == compri

def test_2351_zip():
    L1 = [3, 4, 5]
    L2 = [1, 2, 3]
    L3 = []
    L4 = list(zip(L1, L2))
    for (x1, x2) in L4:
        L3.append(x1+x2)
    logger.info(L3)
    eL3 = [4, 6, 8]
    assert eL3 == L3
def test_2351_zip_map():
    L1 = [3, 4, 5]
    L2 = [1, 2, 3]
    L3 = []
    L3 = list(map(lambda tp : tp[0] + tp[1], (zip(L1, L2))))
    logger.info(L3)
    eL3 = [4, 6, 8]
    assert eL3 == L3

def test_2351_zip_lc():
    L1 = [3, 4, 5]
    L2 = [1, 2, 3]
    L3 = [x1 + x2 for (x1, x2) in list(zip(L1, L2))]
    eL3 = [4, 6, 8]
    assert eL3 == L3

@pytest.mark.skip(reason="Python zip does not have resultSelector")
def test_extra_zip_resultselector():
    pass


# endregion tests

File: csharp/XUnitTests/test_ch23_accum.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Xunit;
using Xunit.Abstractions;
using FluentAssertions;
using FluentAssertions.Equivalency;
using System.Diagnostics.Tracing;
using System.ComponentModel;

namespace XUnitTests
{
    public class test_ch23_accum
    {
        private readonly ITestOutputHelper output; // helper provides interface for WriteLine only, so not logging per se
        public test_ch23_accum(ITestOutputHelper output) { this.output = output; }

        [Fact]
        public void test_2321_map()
        {
            var lst = new List<int> { 2, 5, 9 };
            var lst3 = lst.Select(n => 3 * n);
            var elst3 = new List<int> { 6, 15, 27 };
            lst3.Should().Equal(elst3);
        }// fact
        [Fact]
        public void test_2321_lc()
        {
            var lst = new List<int> { 2, 5, 9 };
            var lst3 = from n in lst select 3 * n;
            var elst3 = new List<int> { 6, 15, 27 };
            lst3.Should().Equal(elst3);
        }// fact
        [Fact(Skip = "C# List of various types not well supported")]
        public void test_2322_map() { }
        [Fact]
        public void test_2323_map()
        {
            var abbrevs = new List<string> { "usa", "esp", "chn", "jpn", "mex", "can", "rus", "rsa", "jam" };
            var abbrevs_upper = abbrevs.Select(s => s.ToUpper());
            var eabbrevs = new List<string> { "USA", "ESP", "CHN", "JPN", "MEX", "CAN", "RUS", "RSA", "JAM" };
            abbrevs_upper.Should().Equal(eabbrevs);
        }
        [Fact]
        public void test_2323_lc()
        {
            var abbrevs = new List<string> { "usa", "esp", "chn", "jpn", "mex", "can", "rus", "rsa", "jam" };
            var abbrevs_upper = from s in abbrevs select s.ToUpper();
            var eabbrevs = new List<string> { "USA", "ESP", "CHN", "JPN", "MEX", "CAN", "RUS", "RSA", "JAM" };
            abbrevs_upper.Should().Equal(eabbrevs);
        }
        [Fact]
        public void test_2331_filter()
        {
            var lst = new List<int> { 3, 4, 6, 7, 0, 1 };
            var evens = lst.Select(n => n).Where(n => n % 2 == 0);
            var elst = new List<int> { 4, 6, 0 };
            evens.Should().Equal(elst);
        }
        [Fact]
        public void test_2331_lc()
        {
            var lst = new List<int> { 3, 4, 6, 7, 0, 1 };
            var evens = from n in lst where (n % 2 == 0) select (n);
            var elst = new List<int> { 4, 6, 0 };
            evens.Should().Equal(elst);
        }
        [Fact]
        public void test_2332_filter()
        {
            var lst_check = new List<string> { "plums", "watermelon", "kiwi", "strawberries", "blueberries", "peaches", "apples", "mangos", "papaya" };
            var filter_testing = lst_check.Where(w => w.Contains("w")).Select(w => w);
            var elst = new List<string> { "watermelon", "kiwi", "strawberries" };
            filter_testing.Should().Equal(elst);
        }
        [Fact]
        public void test_2332_lc()
        {
            var lst_check = new List<string> { "plums", "watermelon", "kiwi", "strawberries", "blueberries", "peaches", "apples", "mangos", "papaya" };
            var filter_testing = from w in lst_check where w.Contains("w") select w;
            var elst = new List<string> { "watermelon", "kiwi", "strawberries" };
            filter_testing.Should().Equal(elst);
        }
        [Fact]
        public void test_2341_mapfilter()
        {
            var things = new List<int> { 3, 4, 6, 7, 0, 1 };
            var lst = things.Where(n => n % 2 == 0).Select(n => n * 2);
            var elst = new List<int> { 8, 12, 0 };
            lst.Should().Equal(elst);
        }
        [Fact]
        public void test_2341_lc()
        {
            var things = new List<int> { 3, 4, 6, 7, 0, 1 };
            var lst = from n in things where n % 2 == 0 select n * 2;
            var elst = new List<int> { 8, 12, 0 };
            lst.Should().Equal(elst);
        }
        [Fact]
        public void test_2342_dict_mapfilter() {
            var tester = new Dictionary<string, List<Dictionary<string, string>>>{
                {"info", new List<Dictionary<string,string>>{
                    new Dictionary<string,string>{ { "name","Lauren" }, { "class standing", "Junior" }, { "major", "Information Science"}},
                    new Dictionary<string,string>{ {"name","Ayo" }, {"class standing","Bachelors" }, {"major", "Information Science"}},
                    new Dictionary<string,string>{ {"name","Kathryn" }, {"class standing","Senior" }, {"major", "Sociology"}},
                    new Dictionary<string,string>{ {"name", "Nick"}, {"class standing", "Junior"}, {"major", "Computer Science"}},
                    new Dictionary<string,string>{ {"name", "Gladys"}, {"class standing", "Sophomore"}, {"major", "History"}},
                    new Dictionary<string,string>{ {"name", "Adam"}, {"major", "Violin Performance"}, {"class standing", "Senior"}}
                }}
            };
            var compri = tester["info"].Where(x => x.ContainsKey("name")).Select(x => x["name"]);
            var ecompri = new List<string> { "Lauren", "Ayo", "Kathryn", "Nick", "Gladys", "Adam"};
            compri.Should().Equal(ecompri);
        }
        [Fact]
        public void test_2342_dict_lc()
        {
            var tester = new Dictionary<string, List<Dictionary<string, string>>>{
                {"info", new List<Dictionary<string,string>>{
                    new Dictionary<string,string>{ { "name","Lauren" }, { "class standing", "Junior" }, { "major", "Information Science"}},
                    new Dictionary<string,string>{ {"name","Ayo" }, {"class standing","Bachelors" }, {"major", "Information Science"}},
                    new Dictionary<string,string>{ {"name","Kathryn" }, {"class standing","Senior" }, {"major", "Sociology"}},
                    new Dictionary<string,string>{ {"name", "Nick"}, {"class standing", "Junior"}, {"major", "Computer Science"}},
                    new Dictionary<string,string>{ {"name", "Gladys"}, {"class standing", "Sophomore"}, {"major", "History"}},
                    new Dictionary<string,string>{ {"name", "Adam"}, {"major", "Violin Performance"}, {"class standing", "Senior"}}
                }}
            };
            var compri = from x in tester["info"] where x.ContainsKey("name") select x["name"];
            var ecompri = new List<string> { "Lauren", "Ayo", "Kathryn", "Nick", "Gladys", "Adam" };
            compri.Should().Equal(ecompri);
        }
        [Fact]
        public void test_2351_zip() {
            var L1 = new List<int> { 3, 4, 5 };
            var L2 = new List<int> { 1, 2, 3 };
            var L3 = new List<int>();
            var L4 = L1.Zip(L2);
            foreach( var (x1, x2)  in L4)
            {
                L3.Add(x1 + x2);
                // L3 = L3.Append(x1 + x2).ToList(); // inefficient to create a new List each time
            }
            var eL3 = new List<int> { 4, 6, 8 };
            L3.Should().Equal(eL3);
        }
        [Fact]
        public void test_2351_zip_map()
        {
            var L1 = new List<int> { 3, 4, 5 };
            var L2 = new List<int> { 1, 2, 3 };
            var L3 = L1.Zip(L2).Select(x => x.First + x.Second);
            var eL3 = new List<int> { 4, 6, 8 };
            L3.Should().Equal(eL3);
        }
        [Fact]
        public void test_2351_zip_lc()
        {
            var L1 = new List<int> { 3, 4, 5 };
            var L2 = new List<int> { 1, 2, 3 };
            var L3 = from x in L1.Zip(L2) select x.First + x.Second;
            var eL3 = new List<int> { 4, 6, 8 };
            L3.Should().Equal(eL3);
        }
        [Fact]
        public void test_extra_zip_resultselector()
        { // Zip takes a third argument that specifies how to put the elements together, Default is an Pair
            int[] numberSequence = { 10, 20, 30, 40, 50 };
            string[] wordSequence = { "Ten", "Twenty", "Thirty", "Forty" };
            var resultSequence = numberSequence.Zip(wordSequence, (first, second) => first + "-" + second);
            string[] expectedSequence = { "10-Ten", "20-Twenty", "30-Thirty", "40-Forty" };
            resultSequence.Should().Equal(expectedSequence);
        }
    }// class
}// namespace
